File: modules/uc-engine/android/src/main/java/expo/modules/ucengine/NativeAnalyticsHost.kt
package expo.modules.ucengine

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.util.AtomicFile
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.time.Instant
import java.util.Locale
import java.util.TimeZone
import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import org.json.JSONArray
import org.json.JSONObject
import uniffi.uc_engine_uniffi.BindingAnalyticsEvent
import uniffi.uc_engine_uniffi.BindingAnalyticsGroupIdentify
import uniffi.uc_engine_uniffi.BindingAnalyticsHost
import uniffi.uc_engine_uniffi.BindingAnalyticsHostException
import uniffi.uc_engine_uniffi.BindingAnalyticsIdentify
import uniffi.uc_engine_uniffi.BindingAnalyticsIdentityChange

private const val ANALYTICS_PREFERENCES = "uc_engine_analytics"
private const val POSTHOG_ENDPOINT = "https://us.i.posthog.com/i/v0/e/"

internal class AndroidPostHogAnalyticsHost(
  context: Context,
  private var appVersion: String,
  preferencesName: String = ANALYTICS_PREFERENCES,
  queueDirectoryOverride: File? = null,
  projectKeyOverride: String? = null,
  private val endpoint: String = POSTHOG_ENDPOINT
) : BindingAnalyticsHost {
  private val applicationContext = context.applicationContext
  private val preferences = applicationContext.getSharedPreferences(preferencesName, Context.MODE_PRIVATE)
  private val queueDirectory = queueDirectoryOverride ?: File(applicationContext.filesDir, "uc-engine-analytics/queue")
  private val executor = Executors.newSingleThreadExecutor()
  private val deliveryRunning = AtomicBoolean(false)
  private val activeDeviceCount = AtomicInteger(0)
  private val queueSequence = AtomicLong(0)
  private val sessionId = UUID.randomUUID().toString()
  private val projectKey = projectKeyOverride ?: runCatching {
    applicationContext.packageManager.getApplicationInfo(
      applicationContext.packageName,
      PackageManager.GET_META_DATA
    ).metaData?.getString("app.uniclipboard.analytics.POSTHOG_PROJECT_KEY")?.trim().orEmpty()
  }.getOrDefault("")

  init {
    if (!queueDirectory.exists() && !queueDirectory.mkdirs()) {
      throw BindingAnalyticsHostException.PersistenceFailed()
    }
    scheduleDelivery()
  }

  fun updateApplicationContext(appVersion: String, activeDeviceCount: Int) {
    this.appVersion = appVersion
    this.activeDeviceCount.set(activeDeviceCount.coerceAtLeast(0))
  }
  fun consentEnabled(): Boolean = preferences.getBoolean("usage_analytics_enabled", true)
  fun setConsentEnabled(enabled: Boolean) {
    if (!preferences.edit().putBoolean("usage_analytics_enabled", enabled).commit()) {
      throw BindingAnalyticsHostException.PersistenceFailed()
    }
    if (enabled) {
      identifyCurrentSpaceGroupIfNeeded()
      scheduleDelivery()
    } else {
      clearPendingEvents()
      persistSpaceGroupIdentified(false)
    }
  }
  fun resetAndIdentify() {
    clearPendingEvents()
    resetTelemetryIdentity()
  }

  fun getAnalyticsState(): Map<String, Any?> {
    val spacePersonId = preferences.getString("space_person_id", null)
    return mapOf(
      "projectKey" to projectKey,
      "consentEnabled" to consentEnabled(),
      "distinctId" to currentDistinctId(),
      "anonymousId" to identifier("anonymous_user_id"),
      "deviceId" to identifier("analytics_device_id"),
      "spaceGroupKey" to preferences.getString("space_id_hash", null),
      "isIdentified" to (spacePersonId != null)
    )
  }

  fun ensureSpaceContext(spaceId: String?, activeDeviceCount: Int) {
    if (spaceId.isNullOrEmpty()) {
      if (!preferences.edit().remove("space_id_hash").remove("space_group_identified").commit()) {
        throw BindingAnalyticsHostException.PersistenceFailed()
      }
      return
    }
    val groupKey = MessageDigest.getInstance("SHA-256").digest(spaceId.toByteArray())
      .take(8).joinToString("") { "%02x".format(it) }
    if (preferences.getString("space_id_hash", null) != groupKey) {
      if (!preferences.edit().putString("space_id_hash", groupKey)
          .putBoolean("space_group_identified", false).commit()) {
        throw BindingAnalyticsHostException.PersistenceFailed()
      }
    }
    identifyCurrentSpaceGroupIfNeeded(activeDeviceCount)
  }

  override fun capture(event: BindingAnalyticsEvent) {
    if (!consentEnabled() || projectKey.isEmpty()) return
    val properties = try { JSONObject(event.propertiesJson) } catch (_: Exception) {
      throw BindingAnalyticsHostException.DeliveryFailed()
    }
    if (containsSensitiveData(properties)) throw BindingAnalyticsHostException.DeliveryFailed()
    commonProperties().forEach { (key, value) -> properties.put(key, value) }
    properties.put("${'$'}insert_id", UUID.randomUUID().toString())
    properties.put("${'$'}device_id", properties.getString("analytics_device_id"))
    properties.put("${'$'}session_id", sessionId)
    properties.put("${'$'}lib", "uniclipboard-mobile")
    properties.put("${'$'}lib_version", appVersion)
    properties.put("${'$'}os", "Android")
    properties.put("${'$'}os_version", Build.VERSION.RELEASE)
    properties.put("${'$'}device_type", "Mobile")
    properties.put("${'$'}geoip_disable", true)
    preferences.getString("space_id_hash", null)?.let { group ->
      properties.put("space_id_hash", group)
      properties.put("${'$'}groups", JSONObject().put("space", group))
    }
    properties.put("${'$'}set", personProperties(properties))
    properties.put("${'$'}set_once", initialProperties(properties))
    enqueue(event.name, currentDistinctId(), properties)
  }

  override fun identify(payload: BindingAnalyticsIdentify) {
    if (!consentEnabled() || projectKey.isEmpty()) return
    val properties = JSONObject()
      .put("${'$'}anon_distinct_id", payload.oldDistinctId)
      .put("${'$'}lib", "uniclipboard-mobile")
      .put("${'$'}geoip_disable", true)
    addJson(payload.setJson, "${'$'}set", properties)
    addJson(payload.setOnceJson, "${'$'}set_once", properties)
    enqueue("${'$'}identify", payload.newDistinctId, properties)
  }

  override fun groupIdentify(payload: BindingAnalyticsGroupIdentify) {
    if (!preferences.edit().putString("space_id_hash", payload.groupKey)
        .putBoolean("space_group_identified", false).commit()) {
      throw BindingAnalyticsHostException.PersistenceFailed()
    }
    if (!consentEnabled() || projectKey.isEmpty()) return
    val properties = JSONObject()
      .put("${'$'}group_type", payload.groupType)
      .put("${'$'}group_key", payload.groupKey)
      .put("${'$'}lib", "uniclipboard-mobile")
      .put("${'$'}geoip_disable", true)
    addJson(payload.setJson, "${'$'}group_set", properties)
    enqueue("${'$'}groupidentify", currentDistinctId(), properties)
    persistSpaceGroupIdentified(true)
  }

  @Synchronized
  override fun adoptSpacePerson(spacePersonId: String): BindingAnalyticsIdentityChange {
    runCatching { UUID.fromString(spacePersonId) }
      .getOrElse { throw BindingAnalyticsHostException.InvalidIdentity() }
    val previous = currentDistinctId()
    if (!preferences.edit().putString("space_person_id", spacePersonId).commit()) {
      throw BindingAnalyticsHostException.PersistenceFailed()
    }
    return BindingAnalyticsIdentityChange(previous, spacePersonId)
  }

  @Synchronized
  override fun releaseSpacePerson(): BindingAnalyticsIdentityChange {
    val previous = currentDistinctId()
    val anonymous = identifier("anonymous_user_id")
    if (!preferences.edit().remove("space_person_id").remove("space_id_hash").commit()) {
      throw BindingAnalyticsHostException.PersistenceFailed()
    }
    return BindingAnalyticsIdentityChange(previous, anonymous)
  }

  override fun currentSpacePersonId(): String? = preferences.getString("space_person_id", null)

  @Synchronized
  override fun resetTelemetryIdentity(): BindingAnalyticsIdentityChange {
    val previous = currentDistinctId()
    val anonymous = UUID.randomUUID().toString()
    if (!preferences.edit()
        .putString("anonymous_user_id", anonymous)
        .putString("analytics_device_id", UUID.randomUUID().toString())
        .remove("space_person_id").remove("space_id_hash").remove("space_group_identified")
        .remove("has_captured_event").commit()) {
      throw BindingAnalyticsHostException.PersistenceFailed()
    }
    return BindingAnalyticsIdentityChange(previous, anonymous)
  }

  @Synchronized
  private fun currentDistinctId(): String =
    preferences.getString("space_person_id", null) ?: identifier("anonymous_user_id")

  @Synchronized
  private fun identifier(key: String): String {
    preferences.getString(key, null)?.let { if (runCatching { UUID.fromString(it) }.isSuccess) return it }
    val created = UUID.randomUUID().toString()
    if (!preferences.edit().putString(key, created).commit()) {
      throw BindingAnalyticsHostException.PersistenceFailed()
    }
    return created
  }

  private fun commonProperties(): Map<String, Any> {
    val firstRun = !preferences.getBoolean("has_captured_event", false)
    if (firstRun) preferences.edit().putBoolean("has_captured_event", true).apply()
    return mapOf(
      "anonymous_user_id" to identifier("anonymous_user_id"),
      "analytics_device_id" to identifier("analytics_device_id"),
      "session_id" to sessionId,
      "app_version" to appVersion,
      "app_channel" to if (applicationContext.packageName.endsWith(".dev")) "development" else "production",
      "os" to "android",
      "os_version" to Build.VERSION.RELEASE,
      "arch" to (Build.SUPPORTED_ABIS.firstOrNull() ?: "other"),
      "locale" to Locale.getDefault().toLanguageTag(),
      "timezone" to TimeZone.getDefault().id,
      "install_source" to "play_store",
      "is_first_run" to firstRun,
      "active_device_count" to activeDeviceCount.get()
    )
  }

  private fun identifyCurrentSpaceGroupIfNeeded(deviceCount: Int = activeDeviceCount.get()) {
    val groupKey = preferences.getString("space_id_hash", null) ?: return
    if (!consentEnabled() || projectKey.isEmpty() ||
      preferences.getBoolean("space_group_identified", false)) return
    groupIdentify(BindingAnalyticsGroupIdentify(
      "space",
      groupKey,
      JSONObject().put("device_count", deviceCount.coerceAtLeast(0)).toString()
    ))
  }

  private fun persistSpaceGroupIdentified(identified: Boolean) {
    if (!preferences.edit().putBoolean("space_group_identified", identified).commit()) {
      throw BindingAnalyticsHostException.PersistenceFailed()
    }
  }

  private fun enqueue(event: String, distinctId: String, properties: JSONObject) {
    if (!properties.has("${'$'}insert_id")) {
      properties.put("${'$'}insert_id", UUID.randomUUID().toString())
    }
    val body = JSONObject().put("api_key", projectKey).put("event", event)
      .put("distinct_id", distinctId).put("properties", properties)
      .put("timestamp", Instant.now().toString())
    val fileName = "%020d-%020d-%s.json".format(
      Locale.ROOT,
      System.currentTimeMillis(),
      queueSequence.incrementAndGet(),
      UUID.randomUUID()
    )
    val target = AtomicFile(File(queueDirectory, fileName))
    try {
      val output = target.startWrite()
      try { output.write(body.toString().toByteArray()); target.finishWrite(output) }
      catch (error: Throwable) { target.failWrite(output); throw error }
      pendingEvents().dropLast(256).forEach(File::delete)
    } catch (_: Exception) { throw BindingAnalyticsHostException.PersistenceFailed() }
    scheduleDelivery()
  }

  private fun scheduleDelivery() {
    if (!consentEnabled() || projectKey.isEmpty() || !deliveryRunning.compareAndSet(false, true)) return
    executor.execute {
      try {
        while (consentEnabled()) {
          val file = pendingEvents().firstOrNull() ?: break
          if (!deliver(file)) break
          file.delete()
        }
      } finally { deliveryRunning.set(false) }
    }
  }

  private fun deliver(file: File): Boolean {
    val connection = (URL(endpoint).openConnection() as HttpURLConnection).apply {
      requestMethod = "POST"; connectTimeout = 10_000; readTimeout = 10_000; doOutput = true
      setRequestProperty("Content-Type", "application/json")
    }
    return try {
      connection.outputStream.use { output -> file.inputStream().use { it.copyTo(output) } }
      connection.responseCode in 200..299
    } catch (_: Exception) { false } finally { connection.disconnect() }
  }

  private fun pendingEvents(): List<File> =
    queueDirectory.listFiles { file -> file.extension == "json" }?.sortedBy { it.name }.orEmpty()
  private fun clearPendingEvents() { pendingEvents().forEach(File::delete) }

  private fun addJson(json: String, key: String, target: JSONObject) {
    val value = try { JSONObject(json) } catch (_: Exception) {
      throw BindingAnalyticsHostException.DeliveryFailed()
    }
    if (containsSensitiveData(value)) throw BindingAnalyticsHostException.DeliveryFailed()
    if (value.length() > 0) target.put(key, value)
  }

  private fun containsSensitiveData(value: Any?, key: String? = null): Boolean {
    val forbidden = setOf("clipboard", "device_name", "display_name", "file_name", "filename", "path", "password", "secret", "token", "invitation_code", "credential")
    key?.lowercase()?.replace('-', '_')?.let {
      if (it in forbidden || it.endsWith("_path") || it.endsWith("_content")) return true
    }
    when (value) {
      is JSONObject -> { val keys = value.keys(); while (keys.hasNext()) { val child = keys.next(); if (containsSensitiveData(value.opt(child), child)) return true } }
      is JSONArray -> for (index in 0 until value.length()) if (containsSensitiveData(value.opt(index))) return true
      is String -> { val text = value.lowercase(); if (text.contains("file://") || text.contains("content://") || text.contains("/users/") || text.contains("/var/mobile/")) return true }
    }
    return false
  }

  private fun personProperties(source: JSONObject): JSONObject {
    val result = JSONObject()
    listOf("app_version", "app_channel", "os", "os_version", "arch", "locale", "timezone", "active_device_count", "space_id_hash")
      .forEach { if (source.has(it)) result.put(it, source.get(it)) }
    return result
  }
  private fun initialProperties(source: JSONObject): JSONObject = JSONObject()
    .put("initial_app_version", source.get("app_version"))
    .put("initial_app_channel", source.get("app_channel"))
    .put("initial_os", source.get("os"))
    .put("initial_install_source", source.get("install_source"))
}

File: modules/uc-engine/ios/NativeAnalyticsHost.swift
import CryptoKit
import Foundation

private enum AnalyticsKey {
  static let consent = "usage_analytics_enabled"
  static let anonymous = "anonymous_user_id"
  static let device = "analytics_device_id"
  static let spacePerson = "space_person_id"
  static let spaceGroup = "space_id_hash"
  static let spaceGroupIdentified = "space_group_identified"
  static let hasCaptured = "has_captured_event"
}

final class AppleAnalyticsStore: @unchecked Sendable {
  private let defaults: UserDefaults
  private let queueDirectory: URL
  private let lock = NSLock()
  private var queueSequence: UInt64 = 0

  init() throws {
    guard let groupID = Self.appGroupID(), let defaults = UserDefaults(suiteName: groupID),
      let container = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupID)
    else { throw BindingAnalyticsHostError.ContextUnavailable }
    self.defaults = defaults
    queueDirectory = container.appendingPathComponent("analytics/queue", isDirectory: true)
    do {
      try FileManager.default.createDirectory(at: queueDirectory, withIntermediateDirectories: true)
    } catch { throw BindingAnalyticsHostError.PersistenceFailed }
  }

  var isEnabled: Bool {
    lock.withAnalyticsLock { defaults.object(forKey: AnalyticsKey.consent) as? Bool ?? true }
  }

  func setEnabled(_ enabled: Bool) throws {
    try lock.withAnalyticsLock {
      defaults.set(enabled, forKey: AnalyticsKey.consent)
      guard defaults.synchronize() else { throw BindingAnalyticsHostError.PersistenceFailed }
    }
  }

  init(defaults: UserDefaults, queueDirectory: URL) throws {
    self.defaults = defaults
    self.queueDirectory = queueDirectory
    do {
      try FileManager.default.createDirectory(at: queueDirectory, withIntermediateDirectories: true)
    } catch { throw BindingAnalyticsHostError.PersistenceFailed }
  }

  func anonymousID() throws -> String { try value(forKey: AnalyticsKey.anonymous) }
  func deviceID() throws -> String { try value(forKey: AnalyticsKey.device) }
  func currentSpacePersonID() -> String? {
    lock.withAnalyticsLock { defaults.string(forKey: AnalyticsKey.spacePerson) }
  }
  func currentSpaceGroupKey() -> String? {
    lock.withAnalyticsLock { defaults.string(forKey: AnalyticsKey.spaceGroup) }
  }
  func isSpaceGroupIdentified() -> Bool {
    lock.withAnalyticsLock { defaults.bool(forKey: AnalyticsKey.spaceGroupIdentified) }
  }
  func currentDistinctID() throws -> String {
    try lock.withAnalyticsLock {
      if let spacePerson = defaults.string(forKey: AnalyticsKey.spacePerson) {
        return spacePerson
      }
      return try identifier(forKey: AnalyticsKey.anonymous)
    }
  }

  func adopt(_ value: String) throws -> BindingAnalyticsIdentityChange {
    guard UUID(uuidString: value) != nil else { throw BindingAnalyticsHostError.InvalidIdentity }
    let previous = try currentDistinctID()
    try save(value, forKey: AnalyticsKey.spacePerson)
    return BindingAnalyticsIdentityChange(previousDistinctId: previous, newDistinctId: value)
  }

  func release() throws -> BindingAnalyticsIdentityChange {
    let previous = try currentDistinctID()
    let anonymous = try anonymousID()
    try lock.withAnalyticsLock {
      defaults.removeObject(forKey: AnalyticsKey.spacePerson)
      defaults.removeObject(forKey: AnalyticsKey.spaceGroup)
      defaults.removeObject(forKey: AnalyticsKey.spaceGroupIdentified)
      guard defaults.synchronize() else { throw BindingAnalyticsHostError.PersistenceFailed }
    }
    return BindingAnalyticsIdentityChange(previousDistinctId: previous, newDistinctId: anonymous)
  }

  func resetTelemetryIdentity() throws -> BindingAnalyticsIdentityChange {
    let previous = try currentDistinctID()
    let anonymous = UUID().uuidString.lowercased()
    try lock.withAnalyticsLock {
      defaults.set(anonymous, forKey: AnalyticsKey.anonymous)
      defaults.set(UUID().uuidString.lowercased(), forKey: AnalyticsKey.device)
      defaults.removeObject(forKey: AnalyticsKey.spacePerson)
      defaults.removeObject(forKey: AnalyticsKey.spaceGroup)
      defaults.removeObject(forKey: AnalyticsKey.spaceGroupIdentified)
      defaults.removeObject(forKey: AnalyticsKey.hasCaptured)
      guard defaults.synchronize() else { throw BindingAnalyticsHostError.PersistenceFailed }
    }
    return BindingAnalyticsIdentityChange(previousDistinctId: previous, newDistinctId: anonymous)
  }

  func setSpaceGroupKey(_ value: String, identified: Bool) throws {
    try lock.withAnalyticsLock {
      defaults.set(value, forKey: AnalyticsKey.spaceGroup)
      defaults.set(identified, forKey: AnalyticsKey.spaceGroupIdentified)
      guard defaults.synchronize() else { throw BindingAnalyticsHostError.PersistenceFailed }
    }
  }

  func setSpaceGroupIdentified(_ identified: Bool) throws {
    try lock.withAnalyticsLock {
      defaults.set(identified, forKey: AnalyticsKey.spaceGroupIdentified)
      guard defaults.synchronize() else { throw BindingAnalyticsHostError.PersistenceFailed }
    }
  }

  func clearSpaceGroup() throws {
    try lock.withAnalyticsLock {
      defaults.removeObject(forKey: AnalyticsKey.spaceGroup)
      defaults.removeObject(forKey: AnalyticsKey.spaceGroupIdentified)
      guard defaults.synchronize() else { throw BindingAnalyticsHostError.PersistenceFailed }
    }
  }

  func consumeFirstRun() -> Bool {
    lock.withAnalyticsLock {
      let first = !defaults.bool(forKey: AnalyticsKey.hasCaptured)
      if first { defaults.set(true, forKey: AnalyticsKey.hasCaptured); defaults.synchronize() }
      return first
    }
  }

  func enqueue(_ data: Data) throws {
    do {
      let fileName = lock.withAnalyticsLock { () -> String in
        queueSequence &+= 1
        let timestamp = UInt64(Date().timeIntervalSince1970 * 1_000_000)
        return String(format: "%020llu-%020llu-%@.json", timestamp, queueSequence, UUID().uuidString)
      }
      try data.write(
        to: queueDirectory.appendingPathComponent(fileName),
        options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication]
      )
      for url in pendingEvents().dropLast(256) { try FileManager.default.removeItem(at: url) }
    } catch { throw BindingAnalyticsHostError.PersistenceFailed }
  }

  func pendingEvents() -> [URL] {
    ((try? FileManager.default.contentsOfDirectory(at: queueDirectory, includingPropertiesForKeys: nil)) ?? [])
      .filter { $0.pathExtension == "json" }
      .sorted { $0.lastPathComponent < $1.lastPathComponent }
  }
  func remove(_ url: URL) { try? FileManager.default.removeItem(at: url) }
  func clearPendingEvents() { pendingEvents().forEach(remove) }

  private func value(forKey key: String) throws -> String {
    try lock.withAnalyticsLock { try identifier(forKey: key) }
  }
  private func identifier(forKey key: String) throws -> String {
    if let existing = defaults.string(forKey: key), UUID(uuidString: existing) != nil { return existing }
    let created = UUID().uuidString.lowercased()
    defaults.set(created, forKey: key)
    guard defaults.synchronize() else { throw BindingAnalyticsHostError.PersistenceFailed }
    return created
  }
  private func save(_ value: String, forKey key: String) throws {
    try lock.withAnalyticsLock {
      defaults.set(value, forKey: key)
      guard defaults.synchronize() else { throw BindingAnalyticsHostError.PersistenceFailed }
    }
  }
  private static func appGroupID() -> String? {
    if let configured = Bundle.main.object(forInfoDictionaryKey: "UCAppGroupIdentifier") as? String,
      !configured.isEmpty { return configured }
    guard var bundleID = Bundle.main.bundleIdentifier else { return nil }
    for suffix in [".Keyboard", ".Share"] where bundleID.hasSuffix(suffix) { bundleID.removeLast(suffix.count) }
    return "group.\(bundleID)"
  }
}

final class ApplePostHogAnalyticsHost: BindingAnalyticsHost, @unchecked Sendable {
  private let store: AppleAnalyticsStore
  private let projectKey: String
  private let endpoint: URL
  private let deliveryQueue = DispatchQueue(label: "app.uniclipboard.analytics.delivery")
  private let contextLock = NSLock()
  private let sessionID = UUID().uuidString.lowercased()
  private var appVersion: String
  private var activeDeviceCount = 0
  private var deliveryRunning = false

  convenience init(appVersion: String) throws {
    try self.init(
      appVersion: appVersion,
      store: AppleAnalyticsStore(),
      projectKey: (Bundle.main.object(forInfoDictionaryKey: "UCPostHogProjectKey") as? String) ?? "",
      endpoint: URL(string: "https://us.i.posthog.com/i/v0/e/")!
    )
  }

  init(appVersion: String, store: AppleAnalyticsStore, projectKey: String, endpoint: URL) throws {
    self.store = store
    self.appVersion = appVersion
    self.projectKey = projectKey.trimmingCharacters(in: .whitespacesAndNewlines)
    self.endpoint = endpoint
    scheduleDelivery()
  }

  func updateApplicationContext(appVersion: String, activeDeviceCount: Int) {
    contextLock.withAnalyticsLock { self.appVersion = appVersion; self.activeDeviceCount = max(0, activeDeviceCount) }
  }
  func consentEnabled() -> Bool { store.isEnabled }
  func setConsentEnabled(_ enabled: Bool) throws {
    try store.setEnabled(enabled)
    if enabled {
      try identifyCurrentSpaceGroupIfNeeded()
      scheduleDelivery()
    } else {
      store.clearPendingEvents()
      try store.setSpaceGroupIdentified(false)
    }
  }
  func resetAndIdentify() throws {
    store.clearPendingEvents()
    _ = try resetTelemetryIdentity()
  }

  func getAnalyticsState() throws -> [String: Any?] {
    let spacePersonID = store.currentSpacePersonID()
    return [
      "projectKey": projectKey,
      "consentEnabled": store.isEnabled,
      "distinctId": try store.currentDistinctID(),
      "anonymousId": try store.anonymousID(),
      "deviceId": try store.deviceID(),
      "spaceGroupKey": store.currentSpaceGroupKey(),
      "isIdentified": spacePersonID != nil,
    ]
  }

  func ensureSpaceContext(spaceID: String?, activeDeviceCount: Int) throws {
    guard let spaceID, !spaceID.isEmpty else {
      try store.clearSpaceGroup()
      return
    }
    let groupKey = SHA256.hash(data: Data(spaceID.utf8)).prefix(8)
      .map { String(format: "%02x", $0) }.joined()
    if store.currentSpaceGroupKey() != groupKey {
      try store.setSpaceGroupKey(groupKey, identified: false)
    }
    try identifyCurrentSpaceGroupIfNeeded(activeDeviceCount: activeDeviceCount)
  }

  func capture(event: BindingAnalyticsEvent) throws {
    guard store.isEnabled, !projectKey.isEmpty else { return }
    guard let data = event.propertiesJson.data(using: .utf8),
      let raw = try JSONSerialization.jsonObject(with: data) as? [String: Any]
    else { throw BindingAnalyticsHostError.DeliveryFailed }
    guard !containsSensitiveData(raw) else { throw BindingAnalyticsHostError.DeliveryFailed }
    var properties = raw
    for (key, value) in try commonProperties() { properties[key] = value }
    properties["$insert_id"] = UUID().uuidString.lowercased()
    properties["$device_id"] = properties["analytics_device_id"]
    properties["$session_id"] = properties["session_id"]
    properties["$lib"] = "uniclipboard-mobile"
    properties["$lib_version"] = properties["app_version"]
    properties["$os"] = "iOS"; properties["$os_version"] = properties["os_version"]
    properties["$device_type"] = "Mobile"; properties["$geoip_disable"] = true
    if let group = store.currentSpaceGroupKey() {
      properties["space_id_hash"] = group; properties["$groups"] = ["space": group]
    }
    properties["$set"] = personProperties(properties)
    properties["$set_once"] = initialProperties(properties)
    try enqueue(event.name, distinctID: try store.currentDistinctID(), properties: properties)
  }

  func identify(payload: BindingAnalyticsIdentify) throws {
    guard store.isEnabled, !projectKey.isEmpty else { return }
    var properties: [String: Any] = ["$anon_distinct_id": payload.oldDistinctId, "$lib": "uniclipboard-mobile", "$geoip_disable": true]
    try addJSON(payload.setJson, key: "$set", to: &properties)
    try addJSON(payload.setOnceJson, key: "$set_once", to: &properties)
    try enqueue("$identify", distinctID: payload.newDistinctId, properties: properties)
  }

  func groupIdentify(payload: BindingAnalyticsGroupIdentify) throws {
    try store.setSpaceGroupKey(payload.groupKey, identified: false)
    guard store.isEnabled, !projectKey.isEmpty else { return }
    var properties: [String: Any] = ["$group_type": payload.groupType, "$group_key": payload.groupKey, "$lib": "uniclipboard-mobile", "$geoip_disable": true]
    try addJSON(payload.setJson, key: "$group_set", to: &properties)
    try enqueue("$groupidentify", distinctID: try store.currentDistinctID(), properties: properties)
    try store.setSpaceGroupIdentified(true)
  }
  func adoptSpacePerson(spacePersonId: String) throws -> BindingAnalyticsIdentityChange { try store.adopt(spacePersonId) }
  func releaseSpacePerson() throws -> BindingAnalyticsIdentityChange { try store.release() }
  func currentSpacePersonId() throws -> String? { store.currentSpacePersonID() }
  func resetTelemetryIdentity() throws -> BindingAnalyticsIdentityChange { try store.resetTelemetryIdentity() }

  private func commonProperties() throws -> [String: Any] {
    let context = contextLock.withAnalyticsLock { (appVersion, activeDeviceCount) }
    return [
      "anonymous_user_id": try store.anonymousID(), "analytics_device_id": try store.deviceID(),
      "session_id": sessionID, "app_version": context.0,
      "app_channel": Bundle.main.bundleIdentifier?.contains(".dev") == true ? "development" : "production",
      "os": "ios", "os_version": ProcessInfo.processInfo.operatingSystemVersionString,
      "arch": Self.architecture, "locale": Locale.current.identifier,
      "timezone": TimeZone.current.identifier, "install_source": "app_store",
      "is_first_run": store.consumeFirstRun(), "active_device_count": context.1,
    ]
  }
  private func identifyCurrentSpaceGroupIfNeeded(activeDeviceCount: Int? = nil) throws {
    guard store.isEnabled, !projectKey.isEmpty, !store.isSpaceGroupIdentified(),
      let groupKey = store.currentSpaceGroupKey()
    else { return }
    let count = activeDeviceCount ?? contextLock.withAnalyticsLock { self.activeDeviceCount }
    try groupIdentify(
      payload: BindingAnalyticsGroupIdentify(
        groupType: "space",
        groupKey: groupKey,
        setJson: "{\"device_count\":\(max(0, count))}"
      )
    )
  }
  private func enqueue(_ event: String, distinctID: String, properties: [String: Any]) throws {
    var properties = properties
    if properties["$insert_id"] == nil { properties["$insert_id"] = UUID().uuidString.lowercased() }
    let body: [String: Any] = ["api_key": projectKey, "event": event, "distinct_id": distinctID, "properties": properties, "timestamp": ISO8601DateFormatter().string(from: Date())]
    guard JSONSerialization.isValidJSONObject(body) else { throw BindingAnalyticsHostError.DeliveryFailed }
    try store.enqueue(try JSONSerialization.data(withJSONObject: body)); scheduleDelivery()
  }
  private func scheduleDelivery() {
    deliveryQueue.async { [self] in guard !deliveryRunning, store.isEnabled, !projectKey.isEmpty else { return }; deliveryRunning = true; deliverNext() }
  }
  private func deliverNext() {
    guard store.isEnabled, let url = store.pendingEvents().first else { deliveryRunning = false; return }
    guard let body = try? Data(contentsOf: url) else { store.remove(url); deliverNext(); return }
    var request = URLRequest(url: endpoint); request.httpMethod = "POST"; request.httpBody = body; request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    URLSession.shared.dataTask(with: request) { [self] _, response, _ in
      deliveryQueue.async { [self] in
        if let status = (response as? HTTPURLResponse)?.statusCode,
          (200..<300).contains(status)
        {
          self.store.remove(url)
          self.deliverNext()
        } else {
          self.deliveryRunning = false
        }
      }
    }.resume()
  }
  private func addJSON(_ json: String, key: String, to properties: inout [String: Any]) throws {
    guard let data = json.data(using: .utf8), let value = try JSONSerialization.jsonObject(with: data) as? [String: Any], !containsSensitiveData(value) else { throw BindingAnalyticsHostError.DeliveryFailed }
    if !value.isEmpty { properties[key] = value }
  }
  private func containsSensitiveData(_ value: Any, key: String? = nil) -> Bool {
    let forbidden = Set(["clipboard", "device_name", "display_name", "file_name", "filename", "path", "password", "secret", "token", "invitation_code", "credential"])
    if let key { let normalized = key.lowercased().replacingOccurrences(of: "-", with: "_"); if forbidden.contains(normalized) || normalized.hasSuffix("_path") || normalized.hasSuffix("_content") { return true } }
    if let object = value as? [String: Any] { return object.contains { containsSensitiveData($0.value, key: $0.key) } }
    if let array = value as? [Any] { return array.contains { containsSensitiveData($0) } }
    if let string = value as? String { let lowered = string.lowercased(); return lowered.contains("file://") || lowered.contains("content://") || lowered.contains("/users/") || lowered.contains("/var/mobile/") }
    return false
  }
  private func personProperties(_ source: [String: Any]) -> [String: Any] {
    let keys = ["app_version", "app_channel", "os", "os_version", "arch", "locale", "timezone", "active_device_count", "space_id_hash"]
    return Dictionary(uniqueKeysWithValues: keys.compactMap { key in source[key].map { (key, $0) } })
  }
  private func initialProperties(_ source: [String: Any]) -> [String: Any] {
    ["initial_app_version": source["app_version"], "initial_app_channel": source["app_channel"], "initial_os": source["os"], "initial_install_source": source["install_source"]].compactMapValues { $0 }
  }
  private static var architecture: String {
    #if arch(arm64)
      "arm64"
    #elseif arch(x86_64)
      "x86_64"
    #else
      "other"
    #endif
  }
}

private extension NSLock {
  func withAnalyticsLock<T>(_ operation: () throws -> T) rethrows -> T { lock(); defer { unlock() }; return try operation() }
}
